timeboost/auctioneer.go

```diff
@@ -0,0 +1,411 @@
+			if pubsub.StreamExists(ctx, a.consumer.StreamName(), a.consumer.RedisClient()) {
+				ready <- struct{}{}
+				readyStream <- struct{}{}
+				return
+			}
+			select {
+			case <-ctx.Done():
+				log.Info("Context done while checking redis stream existance", "error", ctx.Err().Error())
+				return
+			case <-time.After(time.Millisecond * 100):
+			}
+		}
+	})
+	a.StopWaiter.LaunchThread(func(ctx context.Context) {
+		select {
+		case <-ctx.Done():
+			log.Info("Context done while waiting a redis stream to be ready", "error", ctx.Err().Error())
+			return
+		case <-ready: // Wait until the stream exists and start consuming iteratively.
+		}
+		log.Info("Stream exists, now attempting to consume data from it")
+		a.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
+			req, err := a.consumer.Consume(ctx)
+			if err != nil {
+				log.Error("Consuming request", "error", err)
+				return 0
+			}
+			if req == nil {
+				// There's nothing in the queue.
+				return time.Millisecond * 250
+			}
+			// Forward the message over a channel for processing elsewhere in
+			// another thread, so as to not block this consumption thread.
+			a.bidsReceiver <- req.Value
+
+			// We received the message, then we ack with a nil error.
+			if err := a.consumer.SetResult(ctx, req.ID, nil); err != nil {
+				log.Error("Error setting result for request", "id", req.ID, "result", nil, "error", err)
+				return 0
+			}
+			return 0
+		})
+	})
+	a.StopWaiter.LaunchThread(func(ctx context.Context) {
+		for {
+			select {
+			case <-readyStream:
+				log.Trace("At least one stream is ready")
+				return // Don't block Start if at least one of the stream is ready.
+			case <-time.After(a.streamTimeout):
+				log.Error("Waiting for redis streams timed out")
+				return
+			case <-ctx.Done():
+				log.Info("Context done while waiting redis streams to be ready, failed to start")
+				return
+			}
+		}
+	})
+
+	// Bid receiver thread.
+	a.StopWaiter.LaunchThread(func(ctx context.Context) {
+		for {
+			select {
+			case bid := <-a.bidsReceiver:
+				log.Info("Consumed validated bid", "bidder", bid.Bidder, "amount", bid.Amount, "round", bid.Round)
+				a.bidCache.add(JsonValidatedBidToGo(bid))
+				// Persist the validated bid to the database as a non-blocking operation.
+				go a.persistValidatedBid(bid)
+			case <-ctx.Done():
+				log.Info("Context done while waiting redis streams to be ready, failed to start")
+				return
+			}
+		}
+	})
+
+	// Auction resolution thread.
+	a.StopWaiter.LaunchThread(func(ctx context.Context) {
+		ticker := newAuctionCloseTicker(a.roundDuration, a.auctionClosingDuration)
+		go ticker.start()
+		for {
+			select {
+			case <-ctx.Done():
+				log.Error("Context closed, autonomous auctioneer shutting down")
+				return
+			case auctionClosingTime := <-ticker.c:
+				log.Info("New auction closing time reached", "closingTime", auctionClosingTime, "totalBids", a.bidCache.size())
+				if err := a.resolveAuction(ctx); err != nil {
```

Comment:
It should be sufficient to wait a second, and allow the other thread that is consuming and processing bids to do its job during that time. Then, that will be the hard cutoff that we require

Reply:
This could result in missed bids I guess. If there are still any bids in the queue after waiting the 2 seconds time I think probably want to at least know about them and log them. The bidders may be upset that they submit a valid bid on time but because of our infra they didnt get selected.